### ISSUE TITLE
Consider using "conventional commits" for Kata 2.0

### ISSUE THREAD

**Issue by jodh-intel** [kata-containers/kata-containers#195]
# Back story

We currently enforce a particular commit format on *all* commits for all Kata components:

- https://github.com/kata-containers/community/blob/master/CONTRIBUTING.md#patch-format

# Idea

With a small amount of effort, we could update our tooling [1] to enforce ["conventional commits"](https://www.conventionalcommits.org) for Kata 2.0.

I've only skimmed the spec for "conventional commits but our format appears pretty similar already. What we could consider is moving from the curent:

```
subsystem: summary
```

... to:

```
type: subsystem: summary
```

So, rather than,

```
docs: Fix a typo
```

... we'd have:

```
fix: docs: Correct typo
```

... or:

```
feat: docs: Add awesome new user guide.
```

# Why?

Moving to "conventional commits" would allow us to enrich the changelogs we provide for releases and would also "focus the mind" when creating the commit to classify what your change actually means.

# Cost

However, although there are benefits, the cost of doing this would be to potentially increase the barrier to entry to new contributors to the project.

# Pragmatic solution?

As such, even if we *do* decide to adopt conventional commits, we may wish to confine it to particular repositories, or more likely (due to #172) either:

- different parts of this repository.
- different file types (maybe we would only enforce this for commits that changed "code" as opposed to documentation fixes?)

Related: https://github.com/kata-containers/runtime/issues/2486#issuecomment-618573352

/cc @chavafg, @amshinde, @gnawux, @mvincerx.

---

[1] - https://github.com/kata-containers/tests/tree/master/cmd/checkcommits, plus the GitHub services (GitHub Actions, https://github.com/kata-containers/community/wiki/The-SoB-check).

**Comment by jodh-intel** [comment]
:doughnut: <-- for @grahamwhaley :smile:

**Comment by grahamwhaley** [comment]
Hmm. I think I'm not yet convinced. I think my basis comes from a couple of pov:
1) I'm not sure we need more precise release docs. Well, OK, let me rephrase that - *only* if we decide we do need more precise and semi-automated release docs should we move to 'conventional commits' imo.
2) afaik, we are in the process of trying to make the PR submission process easier and less complex, and I suspect this would actually make it more onerous and lift the entry barrier higher. I know we probably need to differentiate between first-time/one-off/drive-by committers and major project contributors who are changing core code and features - but just adding that differentiation itself may raise the complexity of the process and entree as a whole.

**Comment by jodh-intel** [comment]
I think there are two main areas that would benefit from this sort of information:

# Enrich release notes

Making these more easily digestible / "scannable" is a good thing to strive for imho as it should be easy to identify "what changed?" and "why?" easily and quickly. That enriches the information already provided by the semantic versioning guarantees into concrete details.

# Simplify Backporting

  This is arguably way more important than the release notes point. Backporting is a somewhat laborious and paintstaking exercise. But it's absolutely essential work.

_____

Having "a way" to identify bugs, security fixes (sub-class of bugs) and features easily (in other words via automation) will help with both these tasks.

As mentioned, we could limit use of conventional commits to core components (runtime, agent, etc). But, on reflection, maybe we don't need to change the commit format itself. Considering the backport scenario, we normally backport sets of commits (PRs ;-), so that is probably where we care most about correct labelling.

# Labels to the rescue?

We could solve these problems entirely using the appropriate [GitHub labels](https://github.com/kata-containers/tests/tree/master/cmd/github-labels) for each PR

 But although we already have that facility, we're not using it as much as we should. The problem with labels as I see it:

- You can only see and apply a label if you are a member of the GitHub org (which new contributors certainly aren't).
- It therefore requires work on the part of the Rota team *per PR*.
- Crucially, it's easy to forget to maintain labels.

# A *more* pragmatic solution?

Finally getting to the point, I think the *simplest* process would be:

- Assume that *all PRs are bug fixes* unless marked otherwise.
- If a PR is:
  - a feature it must have `[FEATURE]` in the title (essential for backporting PR filtering).
  - a regression it should have `[REGRESSION]` in the title.
  - a security fix it should have `[SECURITY]` in the title.

This simply process extends the concept we already use for do-not-merge (`[DNM]`) / work-in-progress (`[WIP]`) PRs and would:

- Minimise the work involved.
- Allow all users to "label" their work appropriately (with help from the Review Team if necessary).
- Allow us to provide "enriched" changelogs for releases.
- Allow us to simplify backporting.

# How to apply the magic title comments

This is trivial using [`hub`](https://github.com/github/hub). For example, to add a `[FEATURE]` prefix:

```sh
pr="1234"
title=$(hub issue show "$pr"|head -1|sed 's/^# //g')

title="[FEATURE] $title"

hub issue update "$pr" -m "$title"
```

wdyt @amshinde, @chavafg, @mvincerx?

**Comment by chavafg** [comment]
I agree that one of our flaky points currently is backporting. Most of the time, we forget to do those and we spent a lot of time before the release process figuring out what needs to be backported. So if these proposals are helping in this area, I see the value.

This last idea will be less painful for new contributors as making changes to the PR title rather than to the commit message will be easier, but the Review Team needs to be pretty aware of when the title needs to be changed. So I would say we still need some kind of enforcement, but that can/should be part of the Review Team responsibility, so that we don't overwhelm the new contributors with the process. And I think that github labels can help us here. If there is no label `bug fix` and there is no [FEATURE]/[SECURITY]/etc in the PR title, then merging should be blocked.

I was also thinking, could we hit a scenario where a single PR could have more than one commit? and one of the commits is a bug fix and the other something else (e.g. a Feature)?

**Comment by jodh-intel** [comment]
Hi @chavafg. Labels, would indeed be a cleaner and safer method, so I agree we should put the burden on the review team to handle those (since potentially only they have the ability to set labels).

+1 for the idea of blocking PR merging unless the correct labels are in place too - that will help us all focus on remembering to add labels!

As far as enforcing those labels, presumably we can use a GitHub action for this /cc @mvincerx.

You are of course correct that a PR can contain both bug and feature commits. That does make the problem harder to solve. One solution of course is that we mandate that a PR only contains commits the following sets of commits

- A "feature PR": general improvement commits + feature commits.
- A "bug PR": general improvement commits + bug fix commits.
  (same applies for "Regression PRs" and "Security PRs").

That slight change in process may result in a few more PRs being raised, but it should again make backporting even easier and safer since if a PR was labeled "bug fix", it should be safe to backport. The cost of doing this would be requiring the review team to review every commit to ensure that a PR raised as a "bug fix" PR didn't contain any new features for example. That said, I think we are already pretty good at identifying unusual PRs like that already.

# PR merge process

The process could then become:

- PR raiser can optionally annotate the PR title with `[BUG]`, `[FEATURE]`, *etc* as a hint to the review team.
  We should create a GitHub PR template https://github.com/kata-containers/.github/tree/master/.github to ask the raiser to think about whether the PR is a "bug" / "feature" / "improvement" / etc (although we derive this from the GitHub issue if we can get GitHub actions more involved in this process).
- PR is automatically blocked from being merged as it hasn't been "categorised" using labels.
- Review team add appropriate labels to the PR.
- PR can be merged if CI successful.

# Post merge

After a PR has been merged, we could then have a CI job which looks for all "non-feature" recently merged PRs and tries to merge them to the stable branches ("speculative backport testing"). The results of these tests could be made available to the review team to assess the suitability of each PR for backporting.

**Comment by ariel-adam** [comment]
We decided to close this issue since we have enough structure as it is.
